main.go

```diff
@@ -25,514 +25,49 @@ package main
-			return
-		}
-	}
-
-	// Scan existing container for IOC
-	if len(*session.Options.ContainerId) > 0 {
-		session.Log.Info("Scanning container %s for IOC...\n", *session.Options.ContainerId)
-		jsonOutput, err = findIOCInContainer(*session.Options.ContainerId, *session.Options.ContainerNS)
-		if err != nil {
-			core.GetSession().Log.Error("error scanning the container: %s", err)
-			return
-		}
-	}
-
-	if jsonOutput == nil {
-		core.GetSession().Log.Error("set either -local or -image-name flag")
-		return
-	}
-
-	jsonFilename, err := core.GetJsonFilepath()
-	if err != nil {
-		core.GetSession().Log.Error("error while retrieving json output: %s", err)
-		return
-	}
-	if jsonFilename != "" {
-		err = jsonOutput.WriteIOC(jsonFilename)
-		if err != nil {
-			core.GetSession().Log.Error("error while writing IOC: %s", err)
-			return
-		}
-	}
-}
-
-func yaraUpdate(newwg *sync.WaitGroup) {
-	defer newwg.Done()
-	if *session.Options.SocketPath != "" && *session.Options.HttpPort != "" {
-		flag.Parse()
-		fmt.Println("Go Tickers Tutorial")
-		// this creates a new ticker which will
-		// `tick` every 1 second.
-		ticker := time.NewTicker(10 * time.Hour)
-
-		// for every `tick` that our `ticker`
-		// emits, we print `tock`
-		for t := range ticker.C {
-			fmt.Println("Invoked at ", t)
-			err := runYaraUpdate()
-			if err != nil {
-				core.GetSession().Log.Fatal("main: failed to serve: %v", err)
-			}
-		}
-	}
-}
-
-func yaraResults(newwg *sync.WaitGroup) {
-	defer newwg.Done()
-	flag.Parse()
-	err := runYaraUpdate()
-	if err != nil {
-		core.GetSession().Log.Fatal("main: failed to serve: %v", err)
-	}
-	fmt.Println("server inside23 port", *session.Options)
-	core.GetSession().Log.Info("server inside23 port", *session.Options)
-	if *session.Options.SocketPath != "" {
-		//core.GetSession().Log.Info("reached inside server")
-		err := server.RunServer(*session.Options.SocketPath, PLUGIN_NAME)
-		if err != nil {
-			core.GetSession().Log.Fatal("main: failed to serve: %v", err)
-		}
-		//core.GetSession().Log.Info("reached at this point")
-	} else if *session.Options.HttpPort != "" {
-		core.GetSession().Log.Info("server inside port")
-		err := server.RunHttpServer(*session.Options.HttpPort)
-		if err != nil {
-			core.GetSession().Log.Fatal("main: failed to serve through http: %v", err)
-		}
-	} else if *session.Options.StandAloneHttpPort != "" {
-		core.GetSession().Log.Info("server inside port")
-		err := server.RunStandaloneHttpServer(*session.Options.StandAloneHttpPort)
-		if err != nil {
-			core.GetSession().Log.Fatal("main: failed to serve through http: %v", err)
-		}
-	} else {
-		runOnce()
-	}
-
-}
-
-func main() {
 
-	//fmt.Println(" Welcome to concurrency")
 	wg.Add(2)
-	go yaraUpdate(&wg)
-	go yaraResults(&wg)
-	//fmt.Println("Waiting To Finish")
+	// go yaraUpdate(&wg)
+	go runner.ScheduleYaraHunterUpdater(opts, &wg)
+	// go yaraResults(&wg)
+	go runner.StartYaraHunter(opts, config, &wg)
```

Comment:
The typical way to use waitgroups is to have a channel that intercepts ctrl+c, and sends it to the goroutines on a channel so that they can exit gracefully. If not, we can get into situations where ctrl+c does not exit the program !